AssertionError with updater should be caught by LifecycleVisitor

Code that causes error:
~~~
C = Fiber!(Dense(Dense(Element(0.0))), rand(10, 10))
w = Fiber!(Dense(Dense(Element(0.0))), rand(10, 10))
@finch begin 
           for j = _, i = _
               C[i, j] += 1.0
           end
           for j = _, i = _ 
               w[j, i] = C[i, j] 
           end
           for i = _, j = _
              C[j, i] = w[j, i]
           end
       end
~~~

And the resulting error message:
~~~
ERROR: AssertionError: (ctx.modes[head.tns]).kind === updater
Stacktrace:
  [1] lower(root::Finch.FinchNotation.FinchNode, ctx::Finch.LowerJulia, #unused#::Finch.DefaultStyle)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:114
  [2] (::Finch.LowerJulia)(root::Finch.FinchNotation.FinchNode, style::Finch.DefaultStyle)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:34
  [3] (::Finch.LowerJulia)(root::Finch.FinchNotation.FinchNode)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:32
  [4] #127
    @ ~/.julia/dev/Finch/src/lower.jl:130 [inlined]
  [5] #122
    @ ~/.julia/dev/Finch/src/lower.jl:15 [inlined]
  [6] contain(f::Finch.var"#122#123"{Finch.var"#127#132"{Finch.FinchNotation.FinchNode}, Finch.LowerJulia}, ctx::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/environment.jl:56
  [7] contain(f::Finch.var"#127#132"{Finch.FinchNotation.FinchNode}, ctx::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:14
  [8] lower(root::Finch.FinchNotation.FinchNode, ctx::Finch.LowerJulia, #unused#::Finch.DefaultStyle)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:127
--- the last 7 lines are repeated 8 more times ---
 [65] (::Finch.LowerJulia)(root::Finch.FinchNotation.FinchNode, style::Finch.DefaultStyle)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:34
 [66] (::Finch.LowerJulia)(root::Finch.FinchNotation.FinchNode)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:32
 [67] (::Finch.var"#284#288")(ctx_3::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/execute.jl:109
 [68] #122
    @ ~/.julia/dev/Finch/src/lower.jl:15 [inlined]
 [69] contain(f::Finch.var"#122#123"{Finch.var"#284#288", Finch.LowerJulia}, ctx::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/environment.jl:56
 [70] contain(f::Finch.var"#284#288", ctx::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:14
 [71] (::Finch.var"#283#287")(ctx_2::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/execute.jl:108
 [72] (::Finch.var"#122#123"{Finch.var"#283#287", Finch.LowerJulia})(code_2::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:15
 [73] contain(f::Finch.var"#122#123"{Finch.var"#283#287", Finch.LowerJulia}, ctx::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/environment.jl:56
 [74] contain(f::Finch.var"#283#287", ctx::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:14
 [75] lower_global(prgm::Finch.FinchNotation.FinchNode, ctx::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/execute.jl:98
 [76] (::Finch.var"#281#282"{Symbol, DataType})(ctx_2::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/execute.jl:86
 [77] (::Finch.var"#122#123"{Finch.var"#281#282"{Symbol, DataType}, Finch.LowerJulia})(code_2::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:15
 [78] contain(f::Finch.var"#122#123"{Finch.var"#281#282"{Symbol, DataType}, Finch.LowerJulia}, ctx::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/environment.jl:56
 [79] contain(f::Finch.var"#281#282"{Symbol, DataType}, ctx::Finch.LowerJulia)
    @ Finch ~/.julia/dev/Finch/src/lower.jl:14
 [80] #execute_code#280
    @ ~/.julia/dev/Finch/src/execute.jl:83 [inlined]
 [81] execute_code
    @ ~/.julia/dev/Finch/src/execute.jl:82 [inlined]
 [82] (::Finch.var"#276#279"{DataType, DataType})(ctx::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/execute.jl:67
 [83] contain(f::Finch.var"#276#279"{DataType, DataType}, ctx::Finch.JuliaContext)
    @ Finch ~/.julia/dev/Finch/src/environment.jl:56
 [84] macro expansion
    @ ~/.julia/dev/Finch/src/execute.jl:66 [inlined]
 [85] (::Finch.var"#275#278"{DataType, DataType})()
    @ Finch ~/.julia/dev/Finch/src/util.jl:71
 [86] #s339#274
    @ ~/.julia/dev/Finch/src/util.jl:80 [inlined]
 [87] var"#s339#274"(::Any, ex::Any, opts::Any)
    @ Finch ./none:0
 [88] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any})
    @ Core ./boot.jl:602
~~~

--- a/docs/src/benchmark.md
+++ b/docs/src/benchmark.md
@@ -40,10 +40,7 @@ using MatrixDepot
 
 ````
 [ Info: verify download of index files...
-[ Info: creating database file
-[ Info: reading index files
-[ Info: downloading: https://sparse.tamu.edu/files/ss_index.mat
-[ Info: downloading index file https://math.nist.gov/MatrixMarket/matrices.html
+[ Info: reading database
 [ Info: adding metadata...
 [ Info: adding svd data...
 [ Info: writing database
@@ -87,14 +84,14 @@ Benchmark the kernel, ignoring allocation costs for y
 ````
 
 ````
-BenchmarkTools.Trial: 10000 samples with 182 evaluations.
- Range (min … max):  576.929 ns … 837.368 ns  ┊ GC (min … max): 0.00% … 0.00%
- Time  (median):     581.324 ns               ┊ GC (median):    0.00%
- Time  (mean ± σ):   581.753 ns ±   5.906 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%
+BenchmarkTools.Trial: 10000 samples with 211 evaluations.
+ Range (min … max):  355.057 ns … 742.299 ns  ┊ GC (min … max): 0.00% … 0.00%
+ Time  (median):     357.621 ns               ┊ GC (median):    0.00%
+ Time  (mean ± σ):   363.120 ns ±  18.219 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%
 
-        ▁▄█▄                                                     
-  ▂▂▃▄▅▇████▁▅▂▂▁▁▁▁▂▂▁▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▂▁▁▁▁▁▁▁▂▁▁▁▂▁▂▂▂▂▂▂▂▂▂▂ ▃
-  577 ns           Histogram: frequency by time          607 ns <
+  ██▇         ▂▂         ▁ ▁▁▁                                  ▁
+  ████▅▄▅▄▆▄▅▇███▇▅▆▆▅▅████████▇▇▆▆▆▆▅▆▆▆▆▇▆▆▄▇▆▆▆▆▆▅▂▄▄▄▅▄▆▄▄▄ █
+  355 ns        Histogram: log(frequency) by time        441 ns <
 
  Memory estimate: 0 bytes, allocs estimate: 0.
 ````
@@ -116,15 +113,15 @@ end
 ````
 
 ````
-BenchmarkTools.Trial: 10000 samples with 159 evaluations.
- Range (min … max):  644.031 ns … 177.583 μs  ┊ GC (min … max):  0.00% … 99.43%
- Time  (median):     949.063 ns               ┊ GC (median):     0.00%
- Time  (mean ± σ):     1.001 μs ±   5.243 μs  ┊ GC (mean ± σ):  15.71% ±  2.98%
+BenchmarkTools.Trial: 10000 samples with 202 evaluations.
+ Range (min … max):  388.203 ns …  20.209 μs  ┊ GC (min … max): 0.00% … 97.45%
+ Time  (median):     409.035 ns               ┊ GC (median):    0.00%
+ Time  (mean ± σ):   433.924 ns ± 660.317 ns  ┊ GC (mean ± σ):  5.46% ±  3.51%
 
-   █▃                                            ▂▂▃▁            
-  ▄██▅▅█▆▇▅▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▁▁▁▁▂▂▂▁▁▁▁▁▁▁▂▃▄▆██████▇▆▄▄▃▃▃▃▂ ▃
-  644 ns           Histogram: frequency by time         1.05 μs <
+  ▆█▅▂      ▂▄▅▅▅▄▄▃▃▂▂▂▃▃▃▂▂▂▂▁▁ ▁▁▁▁▁   ▁                     ▂
+  ████▇▅▅▄▅▅████████████████████████████████▇███▇▇▇▆▇▆▆▅▆▇▇▆▆▅▅ █
+  388 ns        Histogram: log(frequency) by time        490 ns <
 
- Memory estimate: 624 bytes, allocs estimate: 2.
+ Memory estimate: 608 bytes, allocs estimate: 2.
 ````
 

--- a/docs/src/interactive.ipynb
+++ b/docs/src/interactive.ipynb
@@ -49,11 +49,11 @@
    "file_extension": ".jl",
    "mimetype": "application/julia",
    "name": "julia",
-   "version": "1.6.7"
+   "version": "1.9.2"
   },
   "kernelspec": {
-   "name": "julia-1.6",
-   "display_name": "Julia 1.6.7",
+   "name": "julia-1.9",
+   "display_name": "Julia 1.9.2",
    "language": "julia"
   }
  },

--- a/src/execute.jl
+++ b/src/execute.jl
@@ -102,7 +102,7 @@ function lower_global(prgm, ctx)
                 prgm = enforce_lifecycles(prgm)
                 prgm = dimensionalize!(prgm, ctx_2)
                 prgm = concordize(prgm, ctx_2)
-                #prgm = evaluate_partial(prgm, ctx_2)
+                prgm = evaluate_partial(prgm, ctx_2)
                 prgm = simplify(prgm, ctx_2) #appears necessary
                 prgm = instantiate!(prgm, ctx_2)
                 contain(ctx_2) do ctx_3

--- a/src/semantics.jl
+++ b/src/semantics.jl
@@ -76,7 +76,7 @@ Return an iterator over the properly modified tensors in a finch program
 """
 function getresults(node::FinchNode)
     if node.kind === block
-        return mapreduce(getresults, vcat, node.bodies, init=[])
+        return unique(mapreduce(getresults, vcat, node.bodies, init=[]))
     elseif node.kind === declare || node.kind === thaw
         return [node.tns]
     else

--- a/src/tensors/levels/sparsehashlevels.jl
+++ b/src/tensors/levels/sparsehashlevels.jl
@@ -277,7 +277,7 @@ function freeze_level!(lvl::VirtualSparseHashLevel, ctx::AbstractCompiler, pos_s
     push!(ctx.code.preamble, quote
         resize!($(lvl.ex).srt, length($(lvl.ex).tbl))
         copyto!($(lvl.ex).srt, pairs($(lvl.ex).tbl))
-        sort!($(lvl.ex).srt, by=hashkeycmp)
+        sort!($(lvl.ex).srt, by=$hashkeycmp)
         for $p = 2:($pos_stop + 1)
             $(lvl.ex).ptr[$p] += $(lvl.ex).ptr[$p - 1]
         end

--- a/src/transforms/lifecycle.jl
+++ b/src/transforms/lifecycle.jl
@@ -34,6 +34,7 @@ function enforce_lifecycles(prgm)
     close_scope(prgm, LifecycleVisitor())
 end
 
+#assumes arguments to prgm have been visited already and their uses collected
 function open_stmt(prgm, ctx::LifecycleVisitor)
     for (tns, mode) in ctx.uses
         cur_mode = get(ctx.modes, tns, reader())

--- a/src/util.jl
+++ b/src/util.jl
@@ -214,7 +214,7 @@ Run dead code elimination and constant propagation. `ex` is the target Julia exp
 dataflow(ex) = ex |> striplines |> desugar |> propagate_copies |> mark_dead |> prune_dead |> resugar
 
 isassign(x) = x in Set([:+=, :*=, :&=, :|=, :(=)])
-incs = Dict(:+= => :+, :*= => :*, :&= => :&, :|= => :|)
+incs = Dict(:+= => :+, :-= => :-, :*= => :*, :&= => :&, :|= => :|)
 deincs = Dict(:+ => :+=, :* => :*=, :& => :&=, :| => :|=)
 function ispure(x)
     if x isa Symbol

--- a/test/test_issues.jl
+++ b/test/test_issues.jl
@@ -66,12 +66,12 @@ using CIndices
         function f(a::Float64, b::Float64, c::Float64)
             return a+b+c
         end 
-        struct MyAlgebra <: Finch.AbstractAlgebra end
-        Finch.virtualize(ex, ::Type{MyAlgebra}, ::Finch.JuliaContext) = MyAlgebra()
+        struct MyAlgebra115 <: Finch.AbstractAlgebra end
+        Finch.virtualize(ex, ::Type{MyAlgebra115}, ::Finch.JuliaContext) = MyAlgebra115()
         t = Fiber!(SparseList(SparseList(Element(0.0))))
         B = SparseMatrixCSC([0 0 0 0; -1 -1 -1 -1; -2 -2 -2 -2; -3 -3 -3 -3])
         A = dropdefaults(copyto!(Fiber!(SparseList(SparseList(Element(0.0)))), B))
-        @finch algebra=MyAlgebra() (t .= 0; for j=_, i=_; t[i, j] = f(A[i,j], A[i,j], A[i,j]) end)
+        @finch algebra=MyAlgebra115() (t .= 0; for j=_, i=_; t[i, j] = f(A[i,j], A[i,j], A[i,j]) end)
         @test t == B .* 3
     end
 
@@ -81,14 +81,14 @@ using CIndices
         t = Fiber!(SparseList(SparseList(Element(0.0))))
         B = SparseMatrixCSC([0 0 0 0; -1 -1 -1 -1; -2 -2 -2 -2; -3 -3 -3 -3])
         A = dropdefaults(copyto!(Fiber!(SparseList(SparseList(Element(0.0)))), B))
-        @test_logs (:warn, "Performance Warning: non-concordant traversal of t[i, j] (hint: most arrays prefer column major or first index fast, run in fast mode to ignore this warning)") match_mode=:any @test_throws Finch.FinchProtocolError @finch algebra=MyAlgebra() (t .= 0; for i=_, j=_; t[i, j] = A[i, j] end)
+        @test_logs (:warn, "Performance Warning: non-concordant traversal of t[i, j] (hint: most arrays prefer column major or first index fast, run in fast mode to ignore this warning)") match_mode=:any @test_throws Finch.FinchProtocolError @finch (t .= 0; for i=_, j=_; t[i, j] = A[i, j] end)
     end
 
     let
         t = Fiber!(Dense(SparseList(Element(0.0))))
         B = SparseMatrixCSC([0 0 0 0; -1 -1 -1 -1; -2 -2 -2 -2; -3 -3 -3 -3])
         A = dropdefaults(copyto!(Fiber!(Dense(SparseList(Element(0.0)))), B))
-        @test_logs (:warn, "Performance Warning: non-concordant traversal of t[i, j] (hint: most arrays prefer column major or first index fast, run in fast mode to ignore this warning)") match_mode=:any @test_throws Finch.FinchProtocolError @finch algebra=MyAlgebra() (t .= 0; for i=_, j=_; t[i, j] = A[i, j] end)
+        @test_logs (:warn, "Performance Warning: non-concordant traversal of t[i, j] (hint: most arrays prefer column major or first index fast, run in fast mode to ignore this warning)") match_mode=:any @test_throws Finch.FinchProtocolError @finch (t .= 0; for i=_, j=_; t[i, j] = A[i, j] end)
     end
 
     #https://github.com/willow-ahrens/Finch.jl/issues/129
@@ -302,5 +302,23 @@ using CIndices
             end
         end
         @test C == A * B
+    end
+
+    #https://github.com/willow-ahrens/Finch.jl/issues/284
+    let
+        C = Fiber!(Dense(Dense(Element(0.0))), [1 0; 0 1])
+        w = Fiber!(SparseHash{2}(Element(0.0)), [0 0; 0 0])
+        @finch mode=fastfinch begin 
+            for j = _, i = _
+                C[i, j] += 1
+            end
+            for j = _, i = _ 
+                w[j, i] = C[i, j] 
+            end
+            for i = _, j = _
+                C[j, i] = w[j, i]
+            end
+        end
+        @test C == [2.0 1.0; 1.0 2.0]
     end
 end

